backends/dpdk/dpdkProgram.cpp

```diff
@@ -0,0 +1,282 @@
+  }
+  for (auto state : p->states) {
+    if (state->selectExpression) {
+      if (state->selectExpression->is<IR::SelectExpression>()) {
+        auto select = state->selectExpression->to<IR::SelectExpression>();
+        for (auto pair : select->selectCases) {
+          auto got = degree_map.find(pair->state->path->name);
+          if (got != degree_map.end()) {
+            got->second++;
+          }
+        }
+      } else if (auto path =
+                     state->selectExpression->to<IR::PathExpression>()) {
+        auto got = degree_map.find(path->path->name);
+        if (got != degree_map.end()) {
+          got->second++;
+        }
+      }
+    }
+  }
+  degree_map.erase("start");
+  state_map.erase("start");
+
+  while (stack.size() > 0) {
+    auto state = stack.back();
+    stack.pop_back();
+
+    // the main body
+    auto i = new IR::DpdkLabelStatement("L_" + state.name.toString());
+    add_instr(i);
+    auto c = state.components;
+    for (auto stat : c) {
+      DPDK::ConvertStatementToDpdk h(refmap, typemap, 0, this->collector,
+                                     csum_map);
+      stat->apply(h);
+      for (auto i : h.get_instr())
+        add_instr(i);
+    }
+    if (state.selectExpression) {
+      if (auto e = state.selectExpression->to<IR::SelectExpression>()) {
+        const IR::Expression *switch_var;
+        switch_var = e->select->components[0];
+        for (auto v : e->selectCases) {
+          if (!v->keyset->is<IR::DefaultExpression>()) {
+            add_instr(new IR::DpdkJmpEqualStatement("L_" + v->state->path->name,
+                                                    v->keyset, switch_var));
+          } else {
+            auto i = new IR::DpdkJmpStatement("L_" + v->state->path->name);
+            add_instr(i);
+          }
+        }
+      } else if (auto p = state.selectExpression->to<IR::PathExpression>()) {
+        auto i = new IR::DpdkJmpStatement("L_" + p->path->name);
+        add_instr(i);
+      } else {
+        BUG("P4 Parser switch statement has other situations.");
+      }
+    }
+    // ===========
+    if (state.selectExpression) {
+      if (state.selectExpression->is<IR::SelectExpression>()) {
+        auto select = state.selectExpression->to<IR::SelectExpression>();
+        for (auto pair : select->selectCases) {
+          auto result = degree_map.find(pair->state->toString());
+          if (result != degree_map.end()) {
+            result->second--;
+          }
+        }
+      } else if (state.selectExpression->is<IR::PathExpression>()) {
+        auto got = degree_map.find(state.selectExpression->toString());
+        if (got != degree_map.end()) {
+          got->second--;
+        }
+      }
+    }
+
+    for (auto pair : degree_map) {
+      if (pair.second == 0) {
+        auto result = state_map.find(pair.first);
+        if (result != state_map.end()) {
+          stack.push_back(result->second);
+          degree_map.erase(result->first);
```

Comment:
why do this?

Reply:
@gordonwucn please take a look.

Reply:
In order to make a parser in P4, which is a DAG to a linear form, I would need to use topological sorting. `degree_map` stores the information about how many states depend on other states' execution. This piece of code is to find out the states that, currently, no state depends on, remove them from the map and push them into the stack for code generation.

Reply:
There is already a CFG class for topological sorting, but since you have done it already it's fine.
Does this handle cycles?

Reply:
This does not handle cycles. @hanw To support DAG loop detection, can you add a piece of code like this:
`if(degree_map.size() > 0 and stack.size() == 0) BUG("This parser has a loop");`
and insert here https://github.com/p4lang/p4c/blob/b2d6377aca392e160c5990ac618542a84386a44e/backends/dpdk/dpdkProgram.cpp#L240

Reply:
added.